peps/pep-0815.rst

```diff
@@ -0,0 +1,2281 @@
+        Konstantin Schütze <konstin@mailbox.org>,
+        Ralf Gommers <ralf.gommers@gmail.com>,
+        Andrey Talman <atalman@meta.com>,
+        Charlie Marsh <crmarsh416@gmail.com>,
+        Michael Sarahan <msarahan@gmail.com>,
+        Eli Uriegas <eliuriegas@meta.com>,
+        Barry Warsaw <barry@python.org>,
+        Donald Stufft <donald@stufft.io>
+Discussions-To: Pending
+Status: Draft
+Type: Standards Track
+Topic: Packaging
+Created: 03-Dec-2025
+Post-History:
+
+
+Abstract
+========
+
+The Python wheel packaging format uses
+:doc:`packaging:specifications/platform-compatibility-tags` to specify
+wheel’s supported environments based on Python version, ABI, and
+platform (operating system, architecture, core system libraries).  These
+tags are not able to express features of modern hardware. This is
+particularly challenging for the scientific computing, artificial
+intelligence (AI), machine learning (ML), and high-performance computing
+communities, where packages are often built with specific hardware
+accelerations (e.g., NVIDIA CUDA, AMD ROCm), specialized CPU
+instructions (e.g., AVX512_BF16), or other system dependencies.
+
+This PEP proposes “Wheel Variants,” an extension to the
+:doc:`packaging:specifications/binary-distribution-format`.  This
+extension introduces a mechanism for package maintainers to declare
+multiple build variants for the same package version, while allowing
+installers to automatically select the most appropriate variant based on
+system hardware and software capabilities and characteristics. More
+specifically, it proposes:
+
+- An evolution of the wheel format called **Wheel Variant** that allows
+  wheels to be distinguished by hardware or software attributes.
+
+- A **variant provider plugin** interface allowing to dynamically detect
+  platform attributes and select the most suitable wheel.
+
+The goal is to simplify the user experience to a familiar
+``pip install <package>``, while ensuring optimal performance and
+compatibility. This approach allows seamless package resolution without
+requiring intrusive changes to installers, ensures backward
+compatibility, and minimizes the burden on package maintainers.
+
+
+Definitions
+===========
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”,
+“SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this
+document are to be interpreted as described in :rfc:`2119`.
+
+
+Motivation
+==========
+
+The Python packaging ecosystem has evolved to support increasingly
+diverse computing environments. The current software ecosystem often
+relies on platform specific features to pick which binaries are
+compatible with a particular computer. Unfortunately the current wheel
+format cannot adequately express the features of modern hardware. This
+limitation forces package authors into suboptimal distribution
+strategies and creates friction for users attempting to install
+performance-critical packages.
+
```

Comment:
I think I'd add a paragraph here with a concrete example, i.e., bringing up the fact that packages have to be built for specific CUDA or ROCm versions which cannot be included in the wheel tag. As such, you find yourself needing to distribute many builds with identical versions and wheel tags, targeting very different hardware configurations.

Reply:
Yeah, maybe, but we probably need to rewrite the next paragraph then as well.